core/matrix/dense.cpp

```diff
@@ -476,13 +362,29 @@ void Dense<ValueType>::move_to(Dense<next_precision<ValueType>>* result)
 }
 
 
+template <typename ValueType>
+template <typename IndexType>
+void Dense<ValueType>::convert_impl(Coo<ValueType, IndexType>* result) const
+{
+    auto exec = this->get_executor();
+    const auto num_rows = this->get_size()[0];
+
+    Array<int64> row_ptrs{exec, num_rows + 1};
+    exec->run(dense::make_count_nonzeros_per_row(this, row_ptrs.get_data()));
+    exec->run(dense::make_prefix_sum(row_ptrs.get_data(), num_rows + 1));
+    const auto nnz =
+        exec->copy_val_to_host(row_ptrs.get_const_data() + num_rows);
+    result->resize(this->get_size(), nnz);
```

Comment:
Basically all matrix formats that use a row pointer-like structure (Csr, SparsityCsr, Fbcsr, Sellp) are unsuitable for such a function, since you always need to first fill the row pointers, compute memory requirements from that and fill the newly allocated arrays afterwards. All other formats work decently well with a single function.